crypto/evp/p_lib.c

```diff
@@ -941,10 +990,61 @@ void *evp_pkey_make_provided(EVP_PKEY *pk, OPENSSL_CTX *libctx,
         EVP_PKEY_CTX_free(ctx);
     }
 
-    if (tmp_keymgmt != NULL)
-        keydata =
-            evp_keymgmt_util_export_to_provider(pk, tmp_keymgmt);
+    if (tmp_keymgmt == NULL)
+        goto end;
+
+#ifndef FIPS_MODE
+    if (pk->pkey.ptr != NULL) {
+        size_t i;
+
+        /*
+         * Find our keymgmt in the cache.  If it's present, it means that
+         * export has already been done.  We take token copies of the
+         * cached pointers, to have token success values to return.
+         *
+         * TODO(3.0) Right now, we assume we have ample space.  We will
+         * have to think about a cache aging scheme, though, if |i| indexes
+         * outside the array.
+         */
+        i = evp_keymgmt_util_find_pkey_cache_index(pk, tmp_keymgmt);
+        if (!ossl_assert(i < OSSL_NELEM(pk->pkeys)))
+            goto end;
+        if (pk->pkeys[i].keymgmt != NULL) {
+            tmp_keymgmt = pk->pkeys[i].keymgmt;
```

Comment:
Huh? Surely if `pk->pkeys[i].keymgmt != NULL` then we know `evp_keymgmt_util_find_pkey_cache_index` found a valid index, and therefore this assignment is redundant.

Reply:
Not quite, it will return the index of the cache entry having the keymgmt it was given, *or* the next free entry.  I therefore need to the check so `tmp_keymgmt` doesn't get assigned NULL.

Maybe it would be clearer if I had two separate functions, one specifically designed to find a matching keymgmt, and otherwise an index past the maximum index, and another one specifically designed to find the first free entry.  They will essentially do the exact same thing, but one will tune its result a bit.

Reply:
Maybe I should start with documenting that function :wink:

Reply:
I think you misunderstood my point. As you say:

> Not quite, evp_keymgmt_util_find_pkey_cache_index will return the index of the cache entry having the keymgmt it was given, or the next free entry

The "keymgmt it was given" was `tmp_keymgmt`. If the index of the cache entry is valid then `pk->pkeys[i].keymgmt != NULL`. If that is the case then we know that `pk->pkeys[i].keymgmt` and `tmp_keymgmt` are one and the same thing. So therefore this assignment is redundant, since `tmp_keymgmt` is already equal to `pk->pkeys[i].keymgmt`

Reply:
...

Ok, you do have a point there.  I can't recall why I had this assignment...  must be one of those *incarnations*...

Reply:
Line removed